src/canvas.class.ts

```diff
@@ -1,1530 +1,1670 @@
+      return;
+    }
 
-        var centerTransform;
-
-        if (
-          action === 'scale' ||
-          action === 'scaleX' ||
-          action === 'scaleY' ||
-          action === 'resizing'
-        ) {
-          centerTransform = this.centeredScaling || target.centeredScaling;
-        } else if (action === 'rotate') {
-          centerTransform = this.centeredRotation || target.centeredRotation;
-        }
+    let centerTransform;
+
+    if (
+      action === 'scale' ||
+      action === 'scaleX' ||
+      action === 'scaleY' ||
+      action === 'resizing'
+    ) {
+      centerTransform = this.centeredScaling || target.centeredScaling;
+    } else if (action === 'rotate') {
+      centerTransform = this.centeredRotation || target.centeredRotation;
+    }
 
-        return centerTransform ? !altKey : altKey;
-      },
+    return centerTransform ? !modifierKeyPressed : modifierKeyPressed;
+  }
 
-      /**
-       * should disappear before release 4.0
-       * @private
-       */
-      _getOriginFromCorner: function (target, corner) {
-        var origin = {
-          x: target.originX,
-          y: target.originY,
-        };
-
-        if (corner === 'ml' || corner === 'tl' || corner === 'bl') {
-          origin.x = 'right';
-        } else if (corner === 'mr' || corner === 'tr' || corner === 'br') {
-          origin.x = 'left';
-        }
+  /**
+   * Given the control clicked, determine the origin of the transform.
+   * This is bad because controls can totally have custom names
+   * should disappear before release 4.0
+   * @private
+   * @deprecated
+   */
+  _getOriginFromCorner(
+    target: FabricObject,
+    controlName: string
+  ): { x: TOriginX; y: TOriginY } {
+    const origin = {
+      x: target.originX,
+      y: target.originY,
+    };
+    // is a left control ?
+    if (['ml', 'tl', 'bl'].includes(controlName)) {
+      origin.x = 'right';
+      // is a right control ?
+    } else if (['mr', 'tr', 'br'].includes(controlName)) {
+      origin.x = 'left';
+    }
+    // is a top control ?
+    if (['tl', 'mt', 'tr'].includes(controlName)) {
+      origin.y = 'bottom';
+      // is a bottom control ?
+    } else if (['bl', 'mb', 'br'].includes(controlName)) {
+      origin.y = 'top';
+    }
+    return origin;
+  }
 
-        if (corner === 'tl' || corner === 'mt' || corner === 'tr') {
-          origin.y = 'bottom';
-        } else if (corner === 'bl' || corner === 'mb' || corner === 'br') {
-          origin.y = 'top';
-        }
-        return origin;
-      },
+  /**
+   * @private
+   * @param {Event} e Event object
+   * @param {FaricObject} target
+   */
+  _setupCurrentTransform(
+    e: TPointerEvent,
+    target: FabricObject,
+    alreadySelected: boolean
+  ): void {
+    if (!target) {
+      return;
+    }
+    let pointer = this.getPointer(e);
+    if (target.group) {
+      // transform pointer to target's containing coordinate plane
+      // should we use send point to plane?
```

Comment:
ideally a point could have a tag or something that explains in what plane has been trasnformed the idea is useful, but is hard to put in practice because transform takes a generic matrix. Ma point.transform could at least take an extra parameters when is used with knowledge and that extra parameter could tag either the point type Point<Canvas> Point<Object> or add a string that represent the first 4 numbers of the transform matrix.
There are many things we could do to make geometry a little bit more intuitive even in documentation.

For example is not clear to everyone that image.width and height are in the same scale to cropX and cropY but left and top aren't at all.

Reply:
Good point

Reply:
moved to sendPointToPlane in the other branch